### PR TITLE
feat(core): allow clearing cookies from `jwt()`

### PR DESCRIPTION
**Pull request by iMrDJAi**
## ☕️ Reasoning
Currently there is no way to invalidate session from the `jwt()` callback. This change allows clearing the user cookies when `jwt()` returns ~~a falsy value~~ null.

~~Also, a new provider option has been added, `decodeJWT`, which allows passing the existing token of a user (if any) instead of the default one to the `jwt()` callback after provider's `authorize()`. This allows creating providers for the purpose of checking and invalidating existing user sessions.~~

## 🧢 Checklist

- [x] Documentation
- [ ] Tests
- [ ] Ready to be merged

## 🎫 Affected issues
Fixes: https://github.com/nextauthjs/next-auth/discussions/6325

### CODE REVIEW

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Few comments:

- This should be done against https://github.com/nextauthjs/next-auth/tree/main/packages/core as `next-auth` only receives critical fixes at this point
- Let's allow returning `null` in `jwt` which indicates that we should call `sessionStore.clean()`.
- Looks like you did changes to the `session` callback too, but I don't understand the reason. Could we revert that?

File: packages/next-auth/src/core/routes/callback.ts [hunk 1]
```diff
@@ -37,6 +38,13 @@ export default async function callback(params: {
 
   const useJwtSession = sessionStrategy === "jwt"
 
+  const decodedToken = useJwtSession && decodeJWT
```
Comment (balazsorban44) [inline packages/next-auth/src/core/routes/callback.ts:41]: Why is this necessary?

File: packages/next-auth/src/core/types.ts [hunk 2]
```diff
@@ -15,6 +15,7 @@ import type { CookieSerializeOptions } from "cookie"
 import type { NextApiRequest, NextApiResponse } from "next"
 
 export type Awaitable<T> = T | PromiseLike<T>
+export type Falsy = null|undefined|false
```
Comment (balazsorban44) [inline packages/next-auth/src/core/types.ts:18]: Let's not introduce a new type for this, only check for `null`. The other values don't make sense anyway.

File: packages/next-auth/src/core/routes/callback.ts [hunk 3]
```diff
@@ -29,6 +29,7 @@ export default async function callback(params: {
     jwt,
     events,
     callbacks,
+    decodeJWT,
```
Comment (balazsorban44) [inline packages/next-auth/src/core/routes/callback.ts:32]: Unclear what this tries to accomplish. Could you clarify?

File: packages/next-auth/src/core/types.ts [hunk 4]
```diff
@@ -325,7 +326,7 @@ export interface CallbacksOptions<P = Profile, A = Account> {
     session: Session
     user: User | AdapterUser
     token: JWT
-  }) => Awaitable<Session>
+  }) => Awaitable<Session|Falsy>
   /**
    * This callback is called whenever a JSON Web Token is created (i.e. at sign in)
```
Comment (balazsorban44) [inline packages/next-auth/src/core/types.ts:331]: The comment above needs to be extended that explains, returning `null` will invalidate the JWT session.

File: packages/next-auth/src/core/types.ts [hunk 5]
```diff
@@ -325,7 +326,7 @@ export interface CallbacksOptions<P = Profile, A = Account> {
     session: Session
     user: User | AdapterUser
     token: JWT
-  }) => Awaitable<Session>
+  }) => Awaitable<Session|Falsy>
```
Comment (balazsorban44) [inline packages/next-auth/src/core/types.ts:329]: Not sure why this is necessary

**Comment by iMrDJAi** [comment]
@balazsorban44 Hello, sorry for the delay, working on the changes that you have requested.

Regarding the session callback, clearing the cookies could be done when fetching the session, not sure what else to return from it when the token is ~~falsy~~ `null`:

https://github.com/iMrDJAi/next-auth/blob/27dc1b20c0fa6c0d0e3c15f099ef27d8a132159e/packages/next-auth/src/core/routes/session.ts#L63-L68

Does setting the `response.body` to `undefined` does the job? Or should we avoid calling it at all in this case? What do you think?

Regarding the new option `decodeJWT` (or it could be `decodeExistingJWT`, it certainly needs a clearer name), the point is, if the auth cookies exist while the user is attempting to call a provider, it allows to decode the JWT of that user:

https://github.com/iMrDJAi/next-auth/blob/27dc1b20c0fa6c0d0e3c15f099ef27d8a132159e/packages/next-auth/src/core/routes/callback.ts#L41-L46

Then passing it to `jwt()` instead of `defaultToken`:

https://github.com/iMrDJAi/next-auth/blob/27dc1b20c0fa6c0d0e3c15f099ef27d8a132159e/packages/next-auth/src/core/routes/callback.ts#L41-L46

In this way, we could for example make a provider to sign out users, as we have their pervious state.

**Comment by balazsorban44** [comment]
So basically passing the previous token if I understand correctly. That should be a different PR. Could we split it please and simplify this PR?

**Comment by iMrDJAi** [comment]
@balazsorban44 Alright! one thing left:

> The comment above needs to be extended that explains, returning null will invalidate the JWT session.

Working on it...

**Comment by iMrDJAi** [comment]
@balazsorban44 Before creating a new PR, what should we call that option?

**Comment by balazsorban44** [comment]
don't think it should be an option, just give it as `prevToken` or `currentToken`. 🤔

BTW this will resolve #3946

**Comment by rolanday** [comment]
Just curious regarding ...

> "This should be done against https://github.com/nextauthjs/next-auth/tree/main/packages/core as next-auth only receives critical fixes at this point"

What exactly does "at this point" mean? Is next-auth end-of-life?

**Comment by iMrDJAi** [comment]
@rolanday Auth.js will replace NextAuth.js.

@balazsorban44 Actually that's a very good idea, we can pass both pervious and new tokens! I'll play with it a little bit then I'll create a PR.

File: packages/core/src/types.ts [hunk 6]
```diff
@@ -203,6 +203,11 @@ export interface CallbacksOptions<P = Profile, A = Account> {
    * Its content is forwarded to the `session` callback,
    * where you can control what should be returned to the client.
    * Anything else will be kept inaccessible from the client.
+   * 
+   * Returning `null` will invalidate the JWT session by clearing
+   * the user's cookies. You'll still have to monitor and invalidate
+   * unexpired tokens from future requests yourself to prevent
+   * unauthoried access.
```
Comment (balazsorban44) [inline packages/core/src/types.ts:210]:
```suggestion
   * unauthorized access.
```

**Review by balazsorban44 — APPROVED** [review]
Thanks!

**Comment by balazsorban44** [comment]
@rolanday https://github.com/nextauthjs/next-auth/pull/6132#issue-1505602450

**Pull request merged**